url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/Compare/DataOfPoint.java

```diff
@@ -0,0 +1,198 @@
+		return baos.toByteArray();
+	}
+	
+	public void load(DataInputStream dis){
+		try {
+			pid = dis.readInt();
+		} catch (IOException e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
+		}
+		p.readFromFile(dis);
+		int len;
+		try {
+			len = dis.readInt();
+			delaunayIds = new Long[len];
+			for(int i = 0 ; i < len; i++){
+				delaunayIds[i] = dis.readLong();
+			}
+			signature = DataIO.readString(dis);
+			gf_x.readFromFile(dis);
+			gf_y.readFromFile(dis);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+	}
+	
+	public void loadFromBytes(byte[] buffer) throws IOException{
+		DataInputStream dis = new DataInputStream(new ByteArrayInputStream(buffer));
+		pid = dis.readInt();
+		p.readFromFile(dis);
+		int len = dis.readInt();
+		delaunayIds = new Long[len];
+		for(int i = 0 ; i < len; i++){
+			delaunayIds[i] = dis.readLong();
+		}
+		signature = DataIO.readString(dis);
+		gf_x.readFromFile(dis);
+		gf_y.readFromFile(dis);
+		dis.close();
+	}
+	
+	public String getDigest(){
+		String hashValueOfNei = SecurityUtility.computeGeneralHashValue(delaunayIds);
+		return SecurityUtility.computeGeneralHashValue(new String[]{
+				new Integer(pid).toString(),
+				p.getDigest(),
+				hashValueOfNei
+		});
+	}
+	
+	public String getSignature(){
+		return signature;
+	}
+	
+	public void signWithRSA(RSA rsa){
+		signature = rsa.encrypt(getDigest());
+	}
+	
+	/**
+	 * This is for test.
+	 * @throws IOException 
+	 * */
+	public static void main(String[] args) throws IOException{
+		File file = new File("DataOfPoint.testfile");
+		DataOutputStream dos = new DataOutputStream(new FileOutputStream(file));
+		Point p = new Point(6, 8, 10);
+		p.buildByPaillier();
+		DataOfPoint dp = new DataOfPoint(10, p, new Long[]{(long) 0, (long) 1, (long) 2});
+		byte[] data = dp.writeToBytes();
+		int len = data.length;
+		dos.write(data);
+		DataOfPoint dp2 = new DataOfPoint(12, p, new Long[]{(long) 0, (long) 2, (long) 3});
+		byte[] data2 = dp2.writeToBytes();
+		int len2 = data.length;
+		dos.write(data2);
+		dos.flush();
+		dos.close();
+		
+		RandomAccessFile raf = new RandomAccessFile(file, "r");
+		raf.seek(len);
+		data2 = new byte[len2];
+		raf.read(data2);
+		dp2 = new DataOfPoint(data2);
+//		System.out.println(dp2.getPointId());
+		raf.close();
+		
+		file.delete();
+		
+//		testReadFromFile();
+	}
+	
+	public static void testReadFromFile() throws IOException{
+		ArrayList<long[]> testIndexs = new ArrayList<long[]>();
+		DataInputStream dis = new DataInputStream(new FileInputStream(new File("input/NE.dp.idx")));
+		System.out.println("Read from file:");
+		long id, p, l;
+		while(dis.available() > 0){
+			id = dis.readLong();
+			p = dis.readLong();
+			l = dis.readLong();
+			testIndexs.add(new long[]{id, p, l});
+			System.out.println(id);
+		}
+		dis.close();
+		System.out.println("Begin test:");
+		RandomAccessFile raf = new RandomAccessFile("input/NE.dp.dat", "r");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *raf*. The resource is referenced at lines 185, 187. The resource is closed at line 196. There are other execution paths that do not contain closure statements, for example, when RandomAccessFile.seek throws an exception. Ensure *raf* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).